src/hotspot/share/runtime/threadHeapSampler.cpp

```diff
@@ -34,19 +34,18 @@ uint64_t ThreadHeapSampler::_rnd;
 // Default is 512kb.
 volatile int ThreadHeapSampler::_sampling_interval = 512 * 1024;
 
-// Ordering here is important: _log_table first, _log_table_initialized second.
-double ThreadHeapSampler::_log_table[1 << ThreadHeapSampler::FastLogNumBits] = {};
-
-// Force initialization of the log_table.
-bool ThreadHeapSampler::_log_table_initialized = init_log_table();
-
-bool ThreadHeapSampler::init_log_table() {
-  for (int i = 0; i < (1 << FastLogNumBits); i++) {
-    _log_table[i] = (log(1.0 + static_cast<double>(i+0.5) / (1 << FastLogNumBits))
+template<int N>
+struct FastLogTable {
+  constexpr FastLogTable() : _values() {
+    for (int i = 0; i < N; i++) {
+      _values[i] = (log(1.0 + static_cast<double>(i+0.5) / N)
                     / log(2.0));
+    }
   }
-  return true;
-}
+  double _values[N];
+};
+
+const FastLogTable<ThreadHeapSampler::FastLogCount> ThreadHeapSampler::_log_table;
```

Comment:
To guarantee that `_log_table` is evaluated at C++ compile time, it's best to change the code to

```
constexpr FastLogTable<ThreadHeapSampler::FastLogCount> _log_table;
```

There are 2 reasons:

1. C++ guarantees compile-time evaluation only if the expression is used in a "constexpr context". You can read more from [here](https://isocpp.org/blog/2013/01/when-does-a-constexpr-function-get-evaluated-at-compile-time-stackoverflow).
2. In the future, if the `FastLogTable` constructor is modified in a way that cannot be evaluated at compile time, (e.g., someone removes `constexpr` from the constructor's declaration by mistake, the C++ compiler will catch that and tell you that `_log_table` cannot be `constexpr`.

Unfortunately,  you cannot use `constexpr` in forward declarations, so you should either move the definition of `FastLogTable` to the hpp file, or remove the declaration of `_log_table` from the hpp file.

Reply:
Ok, makes sense.

I went with removing the `_log_table` declaration from the hpp file. I think things came out a bit cleaner as a result, too.